src/Microsoft.VisualStudio.ProjectSystem.Managed.VS/ProjectSystem/VS/Runtimes/MissingSdkRuntimeDetector.cs

```diff
@@ -57,29 +67,26 @@ protected override async Task InitializeCoreAsync(CancellationToken cancellation
             // safe to use within IProjectDynamicLoadComponent.LoadAsync.
             _projectGuid = await _project.UnconfiguredProject.GetProjectGuidAsync();
             _missingSetupComponentRegistrationService.RegisterProjectConfiguration(_projectGuid, _project);
-            _projectFaultHandlerService.Forget(RegisterSdkNetCoreRuntimeNeededInProjectAsync(_project), _project.UnconfiguredProject);
+            _subscription = _projectSubscriptionService.ProjectRuleSource.SourceBlock.LinkToAsyncAction(target: OnProjectChangedAsync, _project.UnconfiguredProject);
         }
 
         private async Task RegisterSdkNetCoreRuntimeNeededInProjectAsync(ConfiguredProject project)
         {
-            if (IsInitialized)
-            {
-                var projectProperties = project.Services.ExportProvider.GetExportedValue<ProjectProperties>();
+            var projectProperties = project.Services.ExportProvider.GetExportedValue<ProjectProperties>();
 
-                ConfigurationGeneral configuration = await projectProperties.GetConfigurationGeneralPropertiesAsync();
+            ConfigurationGeneral configuration = await projectProperties.GetConfigurationGeneralPropertiesAsync();
 
-                string? targetFrameworkIdentifier = await configuration.TargetFrameworkIdentifier.GetDisplayValueAsync();
+            string? targetFrameworkIdentifier = await configuration.TargetFrameworkIdentifier.GetDisplayValueAsync();
 
-                string? targetFrameworkVersion = await configuration.TargetFrameworkVersion.GetDisplayValueAsync();
+            string? targetFrameworkVersion = await configuration.TargetFrameworkVersion.GetDisplayValueAsync();
 
-                if (!string.Equals(targetFrameworkIdentifier, s_netCoreTargetFrameworkIdentifier, StringComparisons.FrameworkIdentifiers) ||
-                    string.IsNullOrEmpty(targetFrameworkVersion))
-                {
-                    targetFrameworkVersion = string.Empty;
-                }
```

Comment:
Rather than using `ProjectProperties` here, you can query the `IProjectSubscriptionUpdate` that's passed through for the property values needed. That'll be faster as it doesn't involve MEF or any async calls. You just walk the snapshot object graph.

From our [docs](https://github.com/dotnet/project-system/blob/main/docs/repo/coding-conventions.md#data):

> DO NOT mix snapshot and "live" project data in the same component.
>
> For example, listening to data flow blocks from IProjectSubscriptionService and then reading properties from ProjectProperties within the callback will lead to inconsistent results. The dataflow represents a "snapshot" of the project from changes in the past, whereas ProjectProperties represents the actual live project. These will not always agree. The same applies to consuming other CPS APIs from within a dataflow block, the majority of them use live data to provide results and hence will return results inconsistent with the snapshot that you are reading in the dataflow.

While you're not strictly combining data from snapshots and `ProjectProperties`, you are using the snapshot system's events to determine when to query `ProjectProperties`, which is subject to the same data races. Better just to go all-in on snapshots here.

Reply:
Good to know, changed.